components/PinchZoom/index.ts

```diff
@@ -0,0 +1,274 @@
+  // support yet.
+  private _positioningEl?: HTMLElement;
+  // Current transform.
+  private _x = 0;
+  private _y = 0;
+  private _scale = 1;
+
+  constructor () {
+    super();
+
+    // Watch for children changes.
+    // Note this won't fire for initial contents,
+    // so _stageElChange is also called in connectedCallback.
+    new MutationObserver(() => this._stageElChange())
+      .observe(this, { childList: true });
+
+    // Watch for pointers
+    const pointerTracker: PointerTracker = new PointerTracker(this, {
+      start: (pointer, event) => {
+        // We only want to track 2 pointers at most
+        if (pointerTracker.currentPointers.length === 2) return false;
+        event.preventDefault();
+        return true;
+      },
+      move: previousPointers => {
+        this._onPointerMove(previousPointers, pointerTracker.currentPointers);
+      }
+    });
+
+    this.addEventListener('wheel', event => this._onWheel(event));
+  }
+
+  connectedCallback () {
+    this._stageElChange();
+  }
+
+  get x () {
+    return this._x;
+  }
+
+  get y () {
+    return this._y;
+  }
+
+  get scale () {
+    return this._scale;
+  }
+
+  /**
+   * Update the stage with a given scale/x/y.
+   */
+  setTransform (scale: number = this._scale, x: number = this._x, y: number = this._y, {
+    allowChangeEvent = false
+  }: {
+    /**
+     * Fire a 'change' event if values are different to current values
+     */
+    allowChangeEvent?: boolean
+  } = {}) {
+    // Bail if there's no change
+    if (
+      scale === this._scale &&
+      x === this._x &&
+      y === this._y
+    ) return;
+
+    this._x = x;
+    this._y = y;
+    this._scale = scale;
+
+    this.style.setProperty('--x', this._x + 'px');
+    this.style.setProperty('--y', this._y + 'px');
+    this.style.setProperty('--scale', this._scale + '');
+
+    if (allowChangeEvent) {
+      const event = new Event('change', { bubbles: true });
+      this.dispatchEvent(event);
+    }
+  }
+
+  /**
+   * Called when the direct children of this element change.
+   * Until we have have shadow dom support across the board, we
+   * require a single element to be the child of <pinch-zoom>, and
+   * that's the element we pan/scale.
+   */
+  private _stageElChange () {
+    this._positioningEl = undefined;
+
+    if (this.children.length === 0) {
+      console.warn('There should be at least one child in <pinch-zoom>.');
+      return;
+    }
+
+    const el = this.children[0];
+
+    if (!(el instanceof HTMLElement)) {
+      console.warn('The first child of <pinch-zoom> must be an HTMLElement.');
+      return;
+    }
+
+    this._positioningEl = el;
+
+    if (this.children.length > 1) {
+      console.warn('<pinch-zoom> must not have more than one child.');
+    }
+
+    this.setTransform();
+  }
+
+  private _onWheel (event: WheelEvent) {
+    event.preventDefault();
+
+    const thisRect = this.getBoundingClientRect();
```

Comment:
Is there any reasonable way we can cache this? Listen to `document.onresize`?

Reply:
From in-person chat: There's no way to know bounds haven't changed after caching.

This is a common problem though, and we should have a think about what we need on the platform to make this easier.

Reply:
A primitive for that would be amazing...

Reply:
@todo maybe try ResizeObserver here?

Reply:
Resize observer is useful in cases where we need to immediately react to a change in layout size. Whereas in this case, we only need it on-demand, and we need the position relative to the viewport which resize observer doesn't give us.

Besides, resize observer is Chrome-only, so it doesn't make sense with our target browser support.

Reply:
(Just for completeness sake: If we had broad RO support, you could use ROs to cache the most recent layout values and not have to worry about forcing sync layout.)